### ISSUE TITLE
25 ms step pulse at the end of a move.

### ISSUE THREAD

**Issue by ashelly** [matthewSorensen/k20-grbl-port#4]
I've got grbl loaded on my Teensy3.1.  It's accepting USB serial commands as expected.  The onboard led is turning on for negative Y moves.  So far so good.  So I put an external led on the Y step pin.  It seems to work correctly - fading up and then down with the move, but then there is a bright flash at the end.  I attached a scope, and I'm seeing a large 25ms pulse after it finishes. The rest of the pulses are around 128us.  

The other thing that seems odd is that the pulse spacing seems irregular (not a smooth increase/decrease).  But I'm not 100% sure that that issue is not an error with my trigger setup.

**Comment by physisisics** [comment]
@ashelly  : The 25 ms pulse sounds like something related to the function st_go_idle()  in stepper.c.  The default is to lock the motors for 25 ms after the move has stopped to ensure the machine has stopped, and then disable the motor driver.  In theory, it shouldn't be toggling the Y-step pin though.  Do you have an invert mask applied to your y-step pin?

As far as the irregular pulse spacing goes, that is pretty typical oscope behavior for a signal changing in time.  The scope triggers and collects a lot of data in a buffer at pretty fast speeds.  When the buffer is full, it displays it on screen.  There is typically a long delay between displaying the data and resetting the trigger to collect more data so that a changing signal tends to look "choppy".  The best way to tell if there is an irregularity in pulse timing would be to increase the oscope time window to the longest time you can while still being able to resolve the pulse.  Either that, or less accurately, hooking it up to some motors and listening for erratic behavior in the frequency.  Hopefully, I'll have my cnc running again this weekend and I can test it out.

**Comment by matthewSorensen** [comment]
This is a known issue, but I haven't tracked down the source. This is the first thing I'll investigate when I have access to a scope tomorrow.

Update - both the flash and noticeably jittery step timing seem to be caused by PIT0 (main stepper ISR) and PIT1 (resets pin state) firing too closely. Perhaps resetting the pin state with the DMA hardware will give more consistent timing?

**Comment by ashelly** [comment]
More info: The pulse at the end is directly related to the idle lock time.  When I change the $7 "Step Idle Delay", the pulse width changes to match.   Also if I turn on STEP_PULSE_DELAY, the glitch goes away.

But I had to make 2 changes to get STEP_PULSE_DELAY to work - one compile error, and the time was not being scaled correctly.  See my branch for the fix.

**Comment by ashelly** [comment]
I believe I know the issue:  `pit0_isr()` calls `stepper_go_idle()` at the end of a move, after it has triggered the last pulse.   go_idle uses a spin loop to perform the idle lock delay, so it prevents the isr from exiting until the delay elapses, and that prevents the `pit1_isr()` frrom executing to toggle off the step bits.   I have a fix in my fork - I use timer2 to do the timing for the idle lock delay.  If you like I'll submit a pull request.

**Comment by physisisics** [comment]
That was my first thought too, but I don't think delay_microseconds() disables the isr for pit1, does it? I just looked at the call to delay_microseconds() defined in util.h.  It uses an inline function definition, but it seems every compiler handles that type of function differently.  Maybe it does block the interrupts?

I don't think just setting a second timer without disabling other functions is the best idea.  We really want to make sure the machine's inertia is stopped before executing another command. I think that perhaps the best solution here would be manually inserting a while loop physically counting the clocks. The pit1_isr() should still fire to disable the steppers during this loop, and GRBL wouldn't try to instantiate any other protocols until the loop exits.  I will compile and test this a little later on tonight.  Right now, I'm going to see if I can get a hold of a scope with a 500MSample buffer to test the jitter during acceleration/deceleration.

**Comment by ashelly** [comment]
It's not that delay_microseconds disables the isr, it's that pit1_isr doesn't get called until pit0_isr exits: they are not re-entrant.     All my tests were  consistent with that result.  I'll dig into the interrrupt controller docs tonight if I have time.

**Comment by matthewSorensen** [comment]
Thanks for looking into this!

delay_microseconds is just a loop counting clocks, not a spinlock. 

From my experiments, the issue is a result of pit_1 not firing until pit0 exits. I tried configuring a DMA request to fire on pit1 and toggle the pins, but the documentation for the DMA system is hard to parse.

We may be able to get pit1 to fire with a higher priority than pit0 - I tired playing with the priority system, but need to really dig into the nvic docs.

Lastly, we could switch to using the Flextimer modules to generate pulses, but I'm not sure if they work for one-shot type stuff.

**Comment by physisisics** [comment]
What if the pit_1 isr set a flag when the step pin fired and set it to zero when it turned off.  Then we could have a while loop just before pit_0 was turned off in the stepper_go_idle() function.  That should work, right?  I don't have time to test it until later tonight.

I tested the step pin firing on the large oscope and it seems to be clean and jitter free. I'll post some stuff later tonight.  Perhaps I didn't have the same conditions as you did @ashelly .  Could you post your $$ settings so I could try to reproduce the jitter?  Thanks.

**Comment by ashelly** [comment]
I really don't like while loops in interrupt handlers. I encourage you to look at  the way I fixed it: https://github.com/ashelly/k20-grbl-port/blob/k20/stepper.c   If you are worried about starting the next block before the motor disable happens, just move the `bit_true(sys.execute,EXEC_CYCLE_STOP);` into the t2 isr.

Regarding jitter - After getting the hang of this scope, I see the jitter is much less than I thought.  But one thing that concerns me a bit is that the timer is restarted by calls to `set_step_events_per_minute`, meaning that any ticks that have elapsed since the start of the t0 isr are not accounted for.  I think maybe one of the flextimers may be better for this purpose, since you can change the interrupt trigger level while the timer is still running.   On the other hand, the even better plan may be to move to 0.9 - the newer algorithm does not need to modify the basic interrupt frequency.

**Comment by physisisics** [comment]
@ashelly : you may be saying this already, but I am far from an expert in programming/ microcontroller logic. I think the big issue is that by default, interrupts are disabled when you enter into an interrupt routine.  So even though the interrupt flag for pit_1 is set, it doesn't trigger an interrupt until after pit0_isr() is finished.  On the avr 328p, you can allow interrupts to fire during a separate interrupt routine.  Is there a way to reenable interrupts during pit0_isr() since pit_1 finishes a long time before go_idle does?  I can't find it in the PIT section of the manual.

@ashelly I like your solution, but even setting bit_true(sys.execute,EXEC_CYCLE_STOP); doesn't necessarily mean that a new line of code won't execute because 25 ms is a long time and EXEC_CYCLE_STOP will reboot the system much faster than 25 ms.  I think in this case, a while loop in the isr is called for. If the cycle_stop reboots grbl and a movement is then issued before the pit2_isr() fires, the motors will become disabled during a move.  At least, as far as I understand the code.

Also, on a 48MHz chip, a few cycles don't amount to much when the stepper frequency is 30kHz.  I wouldn't be too worried about the missed clock cycles in set_step_events_per_minute.  

The more I think about it, the more I think a move to V0.9 is the best option.  It seems that most of the code refactoring for 0.9c has settled down and now would be a good time to do the port to teensy.

**Comment by matthewSorensen** [comment]
I'm not particularly worried about the loss of clock cycles due to resetting the timer - it should be more or less constant across ISR calls, and introduced an error in a perdictable and safe direction, uniformly slowing down feeds. We could implement a simple mechanism to compensate for that - PIT0_LDVAL - PIT0_CVAL0 should be approximately the number of timer ticks since the beginning of the ISR.

**Comment by matthewSorensen** [comment]
Ok, so I think I have a fully functional solution to the issue. I reread the documentation for the K20 NVIC registers, and it turns out that the priority number for each interrupt is the high half of a byte, not the low half. This explains why my initial experiments with ISR prioritization failed...

For this issue at hand, setting pit_0 to a lower priority (...higher numerical value...) than pit_1 allows pit_1 to preempt pit_0, and all of the pin reset logic works. I'll commit the change.

This begs the question of what the proper priority for all other interrupts. Presumably, limit logic and pit_1 should have the highest priority, followed by pit_0, and then all of the USB and serial communication.

@physisisics Yes, I think it may be nearly time for 0.9. I have 1.5 weeks free starting the 19th, so I'll be able to dedicate a good chunk of that.

@ashelly Thanks for your comments in 3fb4b66 and 50afeda1093830460b2925fd0d2b281a138885d5. If you want to file a pull request, I'll accept it. Otherwise, I can just apply the patches manually...
